std/bigint.d

```diff
@@ -49,44 +50,113 @@ private:
+     * Throws:
+     *     $(D ConvException) if the string doesn't represent a valid number
      */
-    this(T : const(char)[])(T s, string file = __FILE__, size_t line = __LINE__)
-        pure
+    this(Range)(Range s) if (
+        isBidirectionalRange!Range &&
+        isSomeChar!(ElementType!Range) &&
+        !isInfinite!Range)
     {
-        if (s.length == 0)
-            throw new ConvException("Can't initialize BigInt with "~
-                                    "empty string", file, line);
+        import std.range : chain, walkLength;
+        import std.algorithm.iteration : filter, filterBidirectional;
+        import std.algorithm.comparison: equal;
+        import std.exception : enforce;
+        import std.conv : ConvException;
+
+        enforce!ConvException(!s.empty, "Can't initialize BigInt with an empty range");
+
+        ElementType!Range[2] buff;
         bool neg = false;
-        if (s[0] == '-') {
-            neg = true;
-            s = s[1..$];
-        } else if (s[0] == '+') {
-            s = s[1..$];
-        }
-        data = 0UL;
         bool ok;
-        assert(isZero());
-        if (s.length > 2 && (s[0..2] == "0x" || s[0..2] == "0X"))
+        auto filtered_range = s.filterBidirectional!(a => a != '_');
+        auto len = filtered_range.save.walkLength;
+        ubyte i;
+
+        data = 0UL;
+
+        // check for signs and if the string is a hex value
+        if (len > 1)
+        {
+            for (i = 0; i < 2; i++)
+            {
+                if (filtered_range.empty)
+                    break;
+
+                auto character = filtered_range.front;
+
+                if (i == 0)
+                {
+                    if (character == '-')
+                    {
+                        neg = true;
+                        filtered_range.popFront;
+                        i--;
+                        continue;
+                    }
+                    else if (character == '+')
+                    {
+                        filtered_range.popFront;
+                        i--;
+                        continue;
+                    }
+                }
+                // sign not in the zero position
+                else if (character == '-' || character == '+')
+                {
+                    throw new ConvException("Not a valid numerical string");
+                }
+
+                buff[i] = character;
+                filtered_range.popFront;
+            }
+        }
+
+        if (buff[].equal("0x") || buff[].equal("0X"))
         {
-            ok = data.fromHexString(s[2..$]);
-        } else {
-            ok = data.fromDecimalString(s);
+            ok = data.fromHexString(filtered_range);
         }
-        if (!ok)
-            throw new ConvException("Invalid digit string", file, line);
+        else
+        {
+            ok = data.fromDecimalString(buff[]
+                .filter!(a => a != ElementType!Range.init)
```

Comment:
`buff[].chain(filtered_range)` produces garbage if one of the values wasn't set.